### PR TITLE
Adding a proposed style guide

### PR DESCRIPTION
**Pull request by matt-gardner**
These are the rules we've generally been following, I'm just putting them into a document that we can reference when questions come up.  Does this seem reasonable?  Anything to change or add?

### CODE REVIEW

File: STYLE.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# AllenNLP Style Guide
+
+Our highest priority in writing code (other than correctness) is that our code be easily readable
```
Comment (schmmd) [inline STYLE.md]: I would reword this to "our highest priority for code style is that..."

**Comment by joelgrus** [comment]
I'm sure I'll think of things to add, but none come immediately to mind.

**Comment by schmmd** [comment]
The guide looks reasonable, but I would call out doc strings as to me it's more important to your goals than the other items.  In general, I prefer to be strict about style guidelines through automated tooling and not by human enforcement.  And I really prefer to avoid style debates ;-)

**Comment by matt-gardner** [comment]
@schmmd, what do you mean by "call out doc strings"?  And yes, I'm happy that pylint does most of the enforcement for us.  Some of this document is just explaining what pylint is doing and why.  If there are ways to get more of this into automated checking, I'd be happy with that.

**Comment by schmmd** [comment]
@matt-gardner I mean specifically call out our expectations around doc strings (not sure what they are called in Python) for classes and methods.  In other words, describe the level of documentation we expect for our code.  I see this as an aspect of code style.

**Comment by matt-gardner** [comment]
@schmmd, the second paragraph does this.  I don't have anything more specific than what's there - it depends a lot on what you're documenting - so I had a hard time trying to modify it to incorporate your suggestion.
